### ISSUE TITLE
Await delays on parallel inserts (despite fast parallel execution on Atlas)

### ISSUE THREAD

**Issue by hardcodet** [Automattic/mongoose#12608]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the performance issue has not already been reported

### Slowed down in version

6.6.5

### Node.js version

16.x

### 🦥 Performance issue

We noticed a business case where we have a great number (a few hundred) of parallel inserts to the same collection, which introduced delays of several seconds.

Initially, I thought it was the slow train issue that @vkarpov15 wrote about, but looking at the Atlas profiler, the inserts are hitting Atlas at the same time, and the whole thing seems to take just a few hundred ms:

![image](https://user-images.githubusercontent.com/873009/198536078-2f166ae6-36d2-422e-b571-3147cd712342.png)

This indicates the following:

- Our parallelization works as expected, and mongoose is immediately submitting all those inserts in a timely fashion (otherwise they wouldn't execute together I guess)
- The actual inserts are fast on Atlas too
- However, it takes multiple seconds for the `create` method to return. I measured the `await` time for each individual `create` call as follows:

![image](https://user-images.githubusercontent.com/873009/198536431-51042c94-9812-4e05-8b11-31db26675b97.png)

The logged delays are usually just a few ms if I insert just a few documents. But it takes much longer in case of our big test batch. It also affects all the inserts - it's not like the first inserts would execute faster (actually the last logged ones are faster with only around 800ms)

![image](https://user-images.githubusercontent.com/873009/198536698-a2fc5a0f-c8fb-47f6-9543-ccf33c264dcf.png)

So to sum up our findings:

- Inserts are sent to Atlas in parallel, and I am awaiting them in my code
- Atlas performs those inserts in parallel too, takes just a few hundred ms
- It takes up to multiple seconds to actually get any of the created documents back from mongoose
- Latency isn't an issue (Atlas and our application are in the same data center)
- I played with pool sizes (10, 50, 100, 500), that didn't improve anything

Any idea what could be happening here?

### Steps to Reproduce

Create a few hundred promises that trigger a `create`. Await them using `Promise.all`.

### Expected Behavior

Given that Atlas indicates that the inserts were done after a few hundred ms, we would like to understand where we are loosing those extra seconds, and if that could be fixed without changing architectural principles on our side (e.g. by rewriting to bulkInsert).

**Comment by vkarpov15** [comment]
I'd recommend trying the following:

1. Increase `minPoolSize` as well as `maxPoolSize`. `poolSize` is equivalent to `maxPoolSize`. Try `minPoolSize: 100, maxPoolSize: 100`
2. Try using one `insertMany()` command, or batch your few hundred `create()` calls into `insertMany()` calls of 100 docs apiece.

Given that MongoDB sees each insert being fast, but you're seeing noticable delays, I suspect what's happening is that you have a relatively small connection pool before, and when the batch of inserts comes in the MongoDB Node driver opens a bunch of new sockets to the MongoDB server to send the additional operations. Opening a new socket can be a slow process, which may explain why you're seeing increased latency.

**Comment by hardcodet** [comment]
Hi @vkarpov15 

I tried the pool size, but that didn't change anything, unfortunately. `insertMany` is something I'd like to avoid if possible (it's a rare corner case, and I'd have to rewrite substantial parts of the infrastructure, especially given that I don't think it's a pooling issue: As you can see from the screenshot, all those inserts are being received and processed by Atlas within a few hundred milliseconds. If it was pooling, we would see those dots spread over multiple seconds, no?

![image](https://user-images.githubusercontent.com/873009/201884982-9a8eefa2-aa45-42d3-a610-3fc9318d5f1e.png)

So basically:

* Atlas receives and processes *all* inserts withing < 0.5 seconds, so pooling can't be the limitation
* At the same time, every `await-ed` mongoose insert takes multiple seconds

**Comment by vkarpov15** [comment]
I don't know enough about Atlas charts to know one way or another. Does setting `minPoolSize` help?
